stan/math/opencl/kernel_generator/reduction_2d.hpp

```diff
@@ -0,0 +1,342 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_REDUCTION_2D_HPP
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_REDUCTION_2D_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/opencl/opencl_context.hpp>
+#include <stan/math/opencl/matrix_cl_view.hpp>
+#include <stan/math/opencl/kernel_generator/type_str.hpp>
+#include <stan/math/opencl/kernel_generator/name_generator.hpp>
+#include <stan/math/opencl/kernel_generator/operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/as_operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/colwise_reduction.hpp>
+#include <stan/math/opencl/kernel_generator/rowwise_reduction.hpp>
+#include <stan/math/opencl/kernel_generator/calc_if.hpp>
+#include <map>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+namespace stan {
+namespace math {
+/** \addtogroup opencl_kernel_generator
+ *  @{
+ */
+
+namespace internal {
+class reduction_2d_base {};
+}  // namespace internal
+
+/**
+ * Represents a two dimensional reduction in kernel generator expressions. So as
+ * to be efficient two dimensional reductions are only done partially. That
+ * means instead of 1 element kernel output can have a few rows and a few
+ * columns that need to be reduced to obtain final result. This can be done in a
+ * separate kernel or after copying to CPU. Also two dimensional reductions can
+ * not be used as arguments to other operations - they can only be evaluated.
```

Comment:
> So as to be efficient two dimensional reductions are only performed on one dimension

That is not true! They are done in two dimensions.

> I'm not sure what this means. like you can't do add(reduce_2d_thing(A), B);?

Exactly. Same holds true for colwise reductions.

Reply:
> That is not true! They are done in two dimensions.

Okay just to make sure I get it now this does a reduce on both dimensions returning back dimensions dependent on the compute units available in the device?

This sounds like a partial reduction then? For `2d_reduce`, I'd expect it to do a reduction on both the rows and columns and return a scalar. How about `partial_reduce`? Since this just partially reduces the matrix down by the number of compute units available

Reply:
Yes. Well, colwise reduction is also partial reduction. I think names 2d and colwise make more sense than calling both of them partial reduction. Exact mechanic of how they work are explained in docs.

Reply:
I think no one will remember what cu stands for. Besides I am against any name that does not make it obvious which reduction operates column wise and which in both dimensions. Both of your suggestions so far would make it confusing which one is which.

Reply:
Done.